File: wrench-app/src/main/java/com/izettle/wrench/di/ViewModelModule.java
package com.izettle.wrench.di;

import com.izettle.wrench.applicationlist.ApplicationViewModel;
import com.izettle.wrench.configurationlist.ConfigurationViewModel;
import com.izettle.wrench.dialogs.booleanvalue.FragmentBooleanValueViewModel;
import com.izettle.wrench.dialogs.enumvalue.FragmentEnumValueViewModel;
import com.izettle.wrench.dialogs.integervalue.FragmentIntegerValueViewModel;
import com.izettle.wrench.dialogs.scope.ScopeFragmentViewModel;
import com.izettle.wrench.dialogs.stringvalue.FragmentStringValueViewModel;
import com.izettle.wrench.oss.detail.OssDetailViewModel;
import com.izettle.wrench.oss.list.OssListViewModel;

import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;

@Module
abstract class ViewModelModule {
    @Binds
    @IntoMap
    @ViewModelKey(ApplicationViewModel.class)
    abstract ViewModel bindApplicationViewModel(ApplicationViewModel applicationViewModel);

    @Binds
    @IntoMap
    @ViewModelKey(ConfigurationViewModel.class)
    abstract ViewModel bindConfigurationViewModel(ConfigurationViewModel configurationViewModel);

    @Binds
    @IntoMap
    @ViewModelKey(FragmentStringValueViewModel.class)
    abstract ViewModel bindFragmentStringValueViewModel(FragmentStringValueViewModel fragmentStringValueViewModel);

    @Binds
    @IntoMap
    @ViewModelKey(ScopeFragmentViewModel.class)
    abstract ViewModel bindScopeFragmentViewModel(ScopeFragmentViewModel scopeFragmentViewModel);

    @Binds
    @IntoMap
    @ViewModelKey(FragmentIntegerValueViewModel.class)
    abstract ViewModel bindFragmentIntegerValueViewModel(FragmentIntegerValueViewModel fragmentIntegerValueViewModel);

    @Binds
    @IntoMap
    @ViewModelKey(FragmentEnumValueViewModel.class)
    abstract ViewModel bindFragmentEnumValueViewModel(FragmentEnumValueViewModel fragmentEnumValueViewModel);

    @Binds
    @IntoMap
    @ViewModelKey(FragmentBooleanValueViewModel.class)
    abstract ViewModel bindFragmentBooleanValueViewModel(FragmentBooleanValueViewModel fragmentBooleanValueViewModel);

    @Binds
    @IntoMap
    @ViewModelKey(OssListViewModel.class)
    abstract ViewModel bindOssViewModel(OssListViewModel ossListViewModel);

    @Binds
    @IntoMap
    @ViewModelKey(OssDetailViewModel.class)
    abstract ViewModel bindOssDetailViewModel(OssDetailViewModel ossDetailViewModel);

    @Binds
    abstract ViewModelProvider.Factory bindViewModelFactory(ViewModelFactory factory);
}

File: wrench-app/src/main/java/com/izettle/wrench/di/ViewModelModule.kt
package com.izettle.wrench.di

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.izettle.wrench.applicationlist.ApplicationViewModel
import com.izettle.wrench.configurationlist.ConfigurationViewModel
import com.izettle.wrench.dialogs.booleanvalue.FragmentBooleanValueViewModel
import com.izettle.wrench.dialogs.enumvalue.FragmentEnumValueViewModel
import com.izettle.wrench.dialogs.integervalue.FragmentIntegerValueViewModel
import com.izettle.wrench.dialogs.scope.ScopeFragmentViewModel
import com.izettle.wrench.dialogs.stringvalue.FragmentStringValueViewModel
import com.izettle.wrench.oss.detail.OssDetailViewModel
import com.izettle.wrench.oss.list.OssListViewModel
import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap

@Module
internal abstract class ViewModelModule {
    @Binds
    @IntoMap
    @ViewModelKey(ApplicationViewModel::class)
    internal abstract fun bindApplicationViewModel(applicationViewModel: ApplicationViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(ConfigurationViewModel::class)
    internal abstract fun bindConfigurationViewModel(configurationViewModel: ConfigurationViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(FragmentStringValueViewModel::class)
    internal abstract fun bindFragmentStringValueViewModel(fragmentStringValueViewModel: FragmentStringValueViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(ScopeFragmentViewModel::class)
    internal abstract fun bindScopeFragmentViewModel(scopeFragmentViewModel: ScopeFragmentViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(FragmentIntegerValueViewModel::class)
    internal abstract fun bindFragmentIntegerValueViewModel(fragmentIntegerValueViewModel: FragmentIntegerValueViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(FragmentEnumValueViewModel::class)
    internal abstract fun bindFragmentEnumValueViewModel(fragmentEnumValueViewModel: FragmentEnumValueViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(FragmentBooleanValueViewModel::class)
    internal abstract fun bindFragmentBooleanValueViewModel(fragmentBooleanValueViewModel: FragmentBooleanValueViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(OssListViewModel::class)
    internal abstract fun bindOssViewModel(ossListViewModel: OssListViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(OssDetailViewModel::class)
    internal abstract fun bindOssDetailViewModel(ossDetailViewModel: OssDetailViewModel): ViewModel

    @Binds
    internal abstract fun bindViewModelFactory(factory: ViewModelFactory): ViewModelProvider.Factory
}
